### PR TITLE
Fix planning of duplicate aggs

### LINKED ISSUE #102083: ESQL: cannot plan multiple aggs with the same expression - rule execution limit reached
```
row x = 1, y = 2, z = 3 | stats  a = count(*), b = count(*) | stats  c = count(*)
```

The logical optimizer does not seem to be able to handle this query:

```
{
    "error": {
        "root_cause": [
            {
                "type": "rule_execution_exception",
                "reason": "Rule execution limit [100] reached"
            }
        ],
        "type": "rule_execution_exception",
        "reason": "Rule execution limit [100] reached"
    },
    "status": 500
}
```

The problem seems to be limited to this corner case scenario where you have two aggs with the same expression in one STATS and another STATS after that.

The problem is not limited to `count(*)`, also the following fails with the same error

```
row x = 1, y = 2, z = 3 | stats  a = max(x), b = max(x) | stats  c = max(a)
```

### PR DESCRIPTION
**Pull request by luigidellaquila**
Fixes https://github.com/elastic/elasticsearch/issues/102083

A query with duplicate aggs

```
row x = 1, y = 2, z = 3 | stats  a = count(*), b = count(*) | stats  c = count(*)
```

was optimized by `ReplaceDuplicateAggWithEval` rule, but then the optimization was reverted by `CombineProjections`, resulting in an infinite loop of the logical plan optimizer.

This PR adds a specific check for this rule, so that the duplicate projections are not re-combined in an agg.

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec
Comment (dnhatn) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:702]: @luigidellaquila You might need to append `#[skip:-8.11.99]` to these two tests since they don't work with 8.11.0 or 8.11.1. Similarly to YAML tests, you can adjust the skipped version after backporting.

**Review by costin — COMMENTED** [review]
Does the issue occur when there's grouping involved?
The fix might work but it doesn't look like it addresses the root problem and it has side effects as the projection doesn't get removed anymore.
Likely the issue is caused by the two rules (ReplaceAggs and CombineProjections).

Also the check for NameIds is fragile - better to use an AttributeSet instead.
Separately, there's an optimization here as a is constant, max(a) can be replaced by a.

I'll try to take a look at it myself tomorrow.

**Comment by luigidellaquila** [comment]
> Does the issue occur when there's grouping involved?

Yes it does, the following queries fail as well
```
row x = 1, y = 2, z = 3 | stats  a = count(*), b = count(*) by x | stats  c = count(*)
row x = 1, y = 2, z = 3 | stats  a = count(*), b = count(*) | stats  c = count(*) by a
row x = 1, y = 2, z = 3 | stats  a = count(*), b = count(*) by x | stats  c = count(*) by a
```

> Likely the issue is caused by the two rules (ReplaceAggs and CombineProjections).

Yes, `CombineProjections` pushes projections into aggs and `ReplaceDuplicateAggWithEval` pulls them back (in a loop)

> I'll try to take a look at it myself tomorrow.

👍 thanks

**Review by costin — APPROVED** [review]
After trying out the issue myself, the solution is fairly simple - move ReplaceDuplicateAggWithEval inside the cleanup batch so it gets executed only _once_ at the end:

```
     var cleanup = new Batch<>("Clean Up", new ReplaceDuplicateAggWithEval(), new ReplaceLimitAndSortAsTopN());
```

The rest of the changes are unnecessary.
This solves the issue and furthermore allows the pruning to occur - the test case in this PR would have to be updated since column a is no longer needed:
```
            | stats a = max(salary), b = max(salary)
            | stats  c = max(b)
```
would become 
```
            | stats  b = max(salary)
            | stats  c = max(b)
```

Approving the PR with the fix above.

**Comment by luigidellaquila** [comment]
> The rest of the changes are unnecessary

I tried it and it seems to work fine for that specific query, but it breaks three other tests (npe on layout).
I'm investigating further

**Comment by luigidellaquila** [comment]
The problem is very similar to what happened [here](https://github.com/elastic/elasticsearch/pull/102199#discussion_r1395475667). 
The inverse layout contains null sets for duplicate aggs (in this case the test failures are for a mix of `median(x)` and `percentile(x, 50)`).
The problem was someway hidden by the combination of `ReplaceDuplicateAggWithEval` and `ReplaceLimitAndSortAsTopN`, but with the proposed fix it gets exposed.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec
Comment (astefan) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:702]: Add the queries discussed during the reviews, as well.
And, also, some similar queries touching real data, not only `row`s.

**Review by astefan — APPROVED** [review]
Together with the additional tests, LGTM.

Also, include @costin's suggestion as is after you merged https://github.com/elastic/elasticsearch/pull/102399. The queries you mentioned [here](https://github.com/elastic/elasticsearch/pull/102165#issuecomment-1819020412) do not fail after the other PR is merged.

**Comment by luigidellaquila** [comment]
Thanks @astefan 

> Also, include @costin's suggestion as is after you merged https://github.com/elastic/elasticsearch/pull/102399. The queries you mentioned https://github.com/elastic/elasticsearch/pull/102165#issuecomment-1819020412 do not fail after the other PR is merged.

indeed, I was waiting for the other PR to be merged to proceed with this one

**Comment by luigidellaquila** [comment]
OK, the fix proposed by Costin seems to work fine, but it seems to have some side effects on LIMIT pushdown (investigating further)


In addition, it unveiled another bug:

```
from employees | stats  a = count(*), b = count(*) | stats  c = count(*)

```
```
       c       
---------------
100            
```

it should be `1`. The reason is

```
[2023-11-21T17:50:47,172][TRACE][o.e.x.e.o.LogicalPlanOptimizer] [runTask-0] Rule optimizer.LogicalPlanOptimizer$PruneColumns applied
Limit[500[INTEGER]]                                                                = Limit[500[INTEGER]]
\_Aggregate[[],[COUNT([2a][KEYWORD]) AS c]]                                        = \_Aggregate[[],[COUNT([2a][KEYWORD]) AS c]]
  \_Aggregate[[],[COUNT([2a][KEYWORD]) AS a, COUNT([2a][KEYWORD]) AS b]]           !   \_EsRelation[employees][avg_worked_seconds{f}#256, birth_date{f}#257, emp_n..]
    \_EsRelation[employees][avg_worked_seconds{f}#256, birth_date{f}#257, emp_n..] ! 
```
that is obviously incorrect (you cannot completely prune an aggregate, it would change the query semantics).

The problem does not seem to be related to this PR though. On `main` that specific query fails, but the following executes and gives the same (wrong) results

```
from employees | stats  a = count(*) | stats  c = count(*)
```

I'll open a separate issue for that.

**Comment by luigidellaquila** [comment]
OK, it was an easy fix: the delayed ReplaceDuplicateAggWithEval dodged limit pushdown optimizations; executing another `PushDownAndCombineLimits` solves the problem.

**Comment by luigidellaquila** [comment]
No need to backport, it was a regression in 8.12 snapshot

**Pull request merged**
